pages/agent/upgrading_to_v3.md.erb

```diff
@@ -0,0 +1,94 @@
+# Upgrading to Buildkite Agent v3
+
+The Buildkite Agent has changed a lot in v3, but we've tried to keep it as backwards compatible as possible. We'll cover what’s changed and how to upgrade to new agents.
+
+<%= toc %>
+
+## What's changed
+
+* [Pipeline Plugins](/docs/agents/plugins) for easily sharing functionality between pipelines
+* The bootstrap (run as a sub-process for every job) has moved from a [shell script](https://github.com/buildkite/agent/blob/2-6-stable/templates/bootstrap.sh) to a sub-command of the `buildkite-agent bootstrap`. This means it's written in golang and is much easier for us to maintain and distribute.
+* Agent meta-data has been renamed to "tags"
+* Built-in [Docker and Docker Compose support](/docs/builds/docker-containerized-builds) has been deprecated and moved to dedicated plugins, [docker-compose](https://github.com/buildkite-plugins/docker-compose-buildkite-plugin) and [docker](https://github.com/buildkite-plugins/docker-buildkite-plugin).
+* Variable interpolation in pipeline.yml
+* Hooks support in Windows
+
+## Bootstrap customizations
+
+If you customized your `bootstrap.sh` file, you will need to move the changes to [hooks documentation](hooks), or update your bootstrap.sh to call `buildkite-agent bootstrap`.
+
+## Docker and Docker Compose support
+
+In v2 we supported a variety of environment variables like `BUILDKITE_DOCKER_COMPOSE_CONTAINER` and `BUILDKITE_DOCKER`. These are deprecated in favour of the [docker-compose](https://github.com/buildkite-plugins/docker-compose-buildkite-plugin) and [docker](https://github.com/buildkite-plugins/docker-buildkite-plugin) pipeline plugin.
+
+You can keep using the old environment variables in v3, but they will be removed in v4.
+
+### Steps using `BUILDKITE_DOCKER_COMPOSE_CONTAINER`
+
+This is a step that uses the v2 `BUILDKITE_DOCKER_COMPOSE_CONTAINER` environment variable to run the command in a docker-compose container:
+
+```yaml
+steps:
+  - label: ':hammer: Tests'
+    command: 'scripts/tests.sh'
+    env:
+      BUILDKITE_DOCKER_COMPOSE_CONTAINER: app
+```
+
+To convert it to the [docker-compose plugin](https://github.com/buildkite-plugins/docker-compose-buildkite-plugin)
+you would do this:
+
+```yaml
+steps:
+  - label: ':hammer: Tests'
+    command: 'scripts/tests.sh'
+    plugins:
+      docker-compose#v1.7.0:
+        run: app
+```
+
+Note that at present these plugins are implemented in bash, so will only work on non-windows platforms. Windows users should continue to use the environment vars.
```

Comment:
Would plugins work if they ran the agent within a "Bash for Windows" shell?

Reply:
Not without some changes, the agent currently checks operating system and uses that to switch between interpreters, batch for windows and shell for linux/unix. We could have an `interpreter` config that let's people make that choice independent of OS, or we could do some smarter detection vs just the OS switch.